src/metalperformanceshaders.cs

```diff
@@ -1135,5 +1528,1894 @@ interface MPSMatrixMultiplication {
+
+		[NullAllowed, Export ("cellGateMemoryWeights", ArgumentSemantic.Retain)]
+		IMPSCnnConvolutionDataSource CellGateMemoryWeights { get; set; }
+
+		[Export ("cellToOutputNeuronType", ArgumentSemantic.Assign)]
+		MPSCnnNeuronType CellToOutputNeuronType { get; set; }
+
+		[Export ("cellToOutputNeuronParamA")]
+		float CellToOutputNeuronParamA { get; set; }
+
+		[Export ("cellToOutputNeuronParamB")]
+		float CellToOutputNeuronParamB { get; set; }
+
+		[Static]
+		[Export ("createLSTMDescriptorWithInputFeatureChannels:outputFeatureChannels:")]
+		MPSLSTMDescriptor CreateLSTMDescriptor (nuint inputFeatureChannels, nuint outputFeatureChannels);
+	}
+
+	[TV (11,0), Mac (10, 13, onlyOn64: true), iOS (11,0)]
+	[BaseType (typeof(MPSState), Name = "MPSRNNRecurrentImageState")]
+	interface MPSRnnRecurrentImageState
+	{
+		[Export ("getRecurrentOutputImageForLayerIndex:")]
+		[return: NullAllowed]
+		MPSImage GetRecurrentOutputImage (nuint layerIndex);
+
+		[Export ("getMemoryCellImageForLayerIndex:")]
+		[return: NullAllowed]
+		MPSImage GetMemoryCellImage (nuint layerIndex);
+	}
+
+	[TV (11,0), Mac (10, 13, onlyOn64: true), iOS (11,0)]
+	[BaseType (typeof(MPSCnnKernel), Name = "MPSRNNImageInferenceLayer")]
+	interface MPSRnnImageInferenceLayer
+	{
+		[Export ("inputFeatureChannels")]
+		nuint InputFeatureChannels { get; }
+
+		[Export ("outputFeatureChannels")]
+		nuint OutputFeatureChannels { get; }
+
+		[Export ("numberOfLayers")]
+		nuint NumberOfLayers { get; }
+
+		[Export ("recurrentOutputIsTemporary")]
+		bool RecurrentOutputIsTemporary { get; set; }
+
+		[Export ("storeAllIntermediateStates")]
+		bool StoreAllIntermediateStates { get; set; }
+
+		[Export ("bidirectionalCombineMode", ArgumentSemantic.Assign)]
+		MPSRnnBidirectionalCombineMode BidirectionalCombineMode { get; set; }
+
+		[Export ("initWithDevice:rnnDescriptor:")]
+		[DesignatedInitializer]
+		IntPtr Constructor (IMTLDevice device, MPSRnnDescriptor rnnDescriptor);
+
+		[Export ("initWithDevice:rnnDescriptors:")]
+		[DesignatedInitializer]
+		IntPtr Constructor (IMTLDevice device, MPSRnnDescriptor[] rnnDescriptors);
+
+		[Export ("encodeSequenceToCommandBuffer:sourceImages:destinationImages:recurrentInputState:recurrentOutputStates:")]
+		void EncodeSequence (IMTLCommandBuffer commandBuffer, MPSImage[] sourceImages, MPSImage[] destinationImages, [NullAllowed] MPSRnnRecurrentImageState recurrentInputState, [NullAllowed] NSMutableArray<MPSRnnRecurrentImageState> recurrentOutputStates);
+
+		[Export ("encodeBidirectionalSequenceToCommandBuffer:sourceSequence:destinationForwardImages:destinationBackwardImages:")]
+		void EncodeBidirectionalSequence (IMTLCommandBuffer commandBuffer, MPSImage[] sourceSequence, MPSImage[] destinationForwardImages, [NullAllowed] MPSImage[] destinationBackwardImages);
+
+		[Export ("initWithCoder:device:")]
+		[DesignatedInitializer]
+		IntPtr Constructor (NSCoder aDecoder, IMTLDevice device);
+
+		[Export ("copyWithZone:device:")]
```

Comment:
I think you might need to add a  `[return: Release()]` attribute here, otherwise we'll leak any return values (since we don't know we get ownership of the object).

It would be best to add a test too, to ensure I'm actually right (should be fairly simple though: just call Copy and dispose the return value; if we over-release it should crash pretty soon).